File: 2665-minimum-time-to-repair-cars/minimum-time-to-repair-cars.py
class Solution:
    def repairCars(self, ranks: List[int], cars: int) -> int:
        def is_possible(time):
            k = 0
            for rank in ranks:
                k += floor(sqrt(time//rank))
                if k >= cars:
                    return True
            return False 

        n = len(ranks)
        left, right = (cars//n) ** 2, max(ranks) * ((ceil(cars/n)) ** 2)
        res = 0
        while left <= right:
            mid = left + (right-left)//2
            if is_possible(mid):
                res = mid
                right = mid - 1
            else:
                left = mid + 1
        return res

File: 2665-minimum-time-to-repair-cars/minimum-time-to-repair-cars.java
class Solution {
    public long repairCars(int[] ranks, int cars) {
        int n = ranks.length;
        long left = (long)Math.pow(cars / n, 2);
        long res = 0;
        for (int rank : ranks) {
            res = Math.max(res, rank);
        }
        long right = res * (long)Math.pow(Math.ceil((double)cars / n), 2);
        
        while (left <= right) {
            long mid = left + (right - left) / 2;
            if (isPossible(mid, ranks, cars)) {
                res = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return res;
    }

    private boolean isPossible(long time, int[] ranks, int cars) {
        long k = 0;
        for (int rank : ranks) {
            k += (long)Math.floor(Math.sqrt((double)time / rank));
            if (k >= cars) {
                return true;
            }
        }
        return false;
    }
}